### PR TITLE
Refactor error with enum codes

### PR DESCRIPTION
**Pull request by fnlctrl**

### CODE REVIEW

File: src/errors-new.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+interface RouterError {
+  type: ErrorTypes
+  message: string
+  from?: RouteLocation
+  to?: RouteLocation
+}
+
+const ErrorTypeMessages = {
```
Comment (posva) [inline src/errors-new.ts]: I think we need some kind of `logError` method, similar to vue-next. In our case error codes are part of the public api but I think the messages should only be logged in dev mode and just print the error in prod
At this point I wonder if `message` should just be kept empty 🤔

File: src/errors-new.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  MatcherLocationNormalized,
+  RouteLocation,
+  RouteLocationNormalized,
+} from './types'
+
+// Using string enums because error codes are exposed to developers
+// and number enums could collide with other error codes in runtime
+export enum ErrorTypes {
+  NO_ROUTE_MATCH = 'NO_ROUTE_MATCH',
+  INVALID_ROUTE_MATCH = 'INVALID_ROUTE_MATCH',
```
Comment (posva) [inline src/errors-new.ts]: This error was removed from the codebase

File: src/errors-new.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  MatcherLocation,
+  MatcherLocationNormalized,
+  RouteLocation,
+  RouteLocationNormalized,
+} from './types'
+
+// Using string enums because error codes are exposed to developers
+// and number enums could collide with other error codes in runtime
+export enum ErrorTypes {
+  NO_ROUTE_MATCH = 'NO_ROUTE_MATCH',
```
Comment (posva) [inline src/errors-new.ts]: Let's put `MATCHER_` in front
```suggestion
  MATCHER_NOT_FOUND = 'MATCHER_NOT_FOUND',
```

File: src/errors-new.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    from: RouteLocationNormalized,
+    to: RouteLocationNormalized
+  ) {
+    return `Navigation cancelled from "${from.fullPath}" to "${to.fullPath}" with a new \`push\` or \`replace\``
+  },
+}
+
+export function createRouterError<Type extends ErrorTypes>(
+  type: Type,
+  ...messageArgs: Parameters<typeof ErrorTypeMessages[Type]>
```
Comment (posva) [inline src/errors-new.ts]: I think we could have a more generic creation error similar to https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgEoHsCukoFEpTpTIQAekIAJgM7L6HEDeAUMm8mAJ4AOEAXMgDkcQcgA+QgEaDmAX2bNQOeEjoEiAQRLkIVWhmzR6RZC3Yce-ISNbs4A6mCigA5nIVLoKlMagAhbQoaNCwcX1NbNi5eAUFpSORJBydXd2YYTBAEMGB0EGQEKAg4SF8AHlxA3WCDMPUoAD4ACmirXABtQVbBAF0AGmRuOCg4AFtqAQB5UeAwCoGuy1EJAGsITnQYNQYGgEoBSrN2ABsIMBJkAF5kEAgAd22iJt3kOFpcBXMYEybT87XOMhQINhmNqC8IO0Aa9aADNnQelcQSNxlD1jDkHCtq14UMUdQep92EUwJgoPkIGk-hdroViqV6hV6hpmsJBANGK9YgALdaiWS7ZhAA

We could remove the stack entry added by the function call too, similarly to https://github.com/vuejs/vue-router/pull/3047/files#diff-b9b97e885420591ac8b317a611558999R6-R8. Although it's better to check if `stack` exists first. But given that the `stack` property is still not considered stable and the benefit is only removing one stack entry, we don't really need to do it

File: src/errors-new.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    type === ErrorTypes.NAVIGATION_CANCELLED ||
+    type === ErrorTypes.NAVIGATION_GUARD_REDIRECT
+  ) {
+    error.from = messageArgs[0]
+    error.to = messageArgs[1]
+  }
+  return error
+}
+
+const propertiesToLog: (keyof RouteLocationNormalized)[] = [
```
Comment (posva) [inline src/errors-new.ts]: I think we can remove the type here and let TS infer it
```suggestion
const propertiesToLog = [
```

File: src/errors-new.ts (hunk 1 above)
Comment (fnlctrl) [reply src/errors-new.ts]: I added a __DEV__ check to remove it in prod build. `logError` seems unnecessary as we're not catching all errors then print/rethrow them like in core, maybe?

File: src/errors-new.ts (hunk 4 above)
Comment (fnlctrl) [reply src/errors-new.ts]: I tried to remove the necessity to specify generic when calling, and ended up with this: https://github.com/vuejs/vue-router-next/blob/ef17a9f75fb1fddb9a0bf946e426079e7c25592b/src/errors-new.ts#L36-L40

File: src/utils/guardToPromiseFn.ts [hunk 6]
```diff
@@ -19,9 +19,15 @@ export function guardToPromiseFn(
...
         // TODO: handle callback
-        if (valid === false) reject(new NavigationAborted(to, from))
+        if (valid === false)
+          reject(createRouterError(ErrorTypes.NAVIGATION_ABORTED, { from, to }))
         else if (isRouteLocation(valid)) {
-          reject(new NavigationGuardRedirect(to, valid))
+          reject(
+            createRouterError(ErrorTypes.NAVIGATION_GUARD_REDIRECT, {
+              from: to,
+              to: valid as RouteLocationNormalized, // todo: remove as
```
Comment (fnlctrl) [inline src/utils/guardToPromiseFn.ts]: here `valid` is just `RouteLocation`, how should we make it normalized?

File: src/router.ts [hunk 7]
```diff
@@ -345,7 +351,13 @@ export function createRouter({
   ) {
     // a more recent navigation took place
     if (pendingLocation !== toLocation) {
-      return triggerError(new NavigationCancelled(toLocation, from), isPush)
+      return triggerError(
+        createRouterError(ErrorTypes.NAVIGATION_CANCELLED, {
+          from: toLocation,
```
Comment (posva) [inline src/router.ts]: isn't it the other way around? `from, to: toLocation`?

File: src/router.ts [hunk 8]
```diff
@@ -397,19 +409,25 @@ export function createRouter({
...
-      if (NavigationGuardRedirect.is(error)) {
+      if (error.type === ErrorTypes.NAVIGATION_GUARD_REDIRECT) {
         // TODO: refactor the duplication of new NavigationCancelled by
         // checking instanceof NavigationError (it's another TODO)
         // a more recent navigation took place
         if (pendingLocation !== toLocation) {
-          return triggerError(new NavigationCancelled(toLocation, from), false)
+          return triggerError(
+            createRouterError(ErrorTypes.NAVIGATION_CANCELLED, {
+              from: toLocation,
```
Comment (posva) [inline src/router.ts]: same here?

File: src/utils/guardToPromiseFn.ts (hunk 6 above)
Comment (posva) [reply src/utils/guardToPromiseFn.ts]: We don't need to make it normalised for the redirect. In that case the `to` property can either be `RouteLocation` or `RouteLocationNormalized`. I forgot about it in my previous review 😅

File: src/errors-new.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+  to: RouteLocationNormalized
+}
+
+type InferErrorType<Type extends ErrorTypes> = Type extends MatcherError['type']
+  ? MatcherError
+  : Type extends NavigationError['type']
+  ? NavigationError
+  : never
+
+const ErrorTypeMessages = __DEV__
```
[the remark this answers is not in the document]
Comment (fnlctrl) [reply src/errors-new.ts]: I tried using the object and only check inside createRouterError, but somehow the compressor didn't strip it out...

[the remark this answers is not in the document]
Comment (posva) [reply src/errors-new.ts]: that's weird, it seems to be like it should like in vue-next. I will take a look

File: src/router.ts (hunk 7 above)
Comment (fnlctrl) [reply src/router.ts]: oops😂

File: src/errors-new.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  currentLocation?: MatcherLocationNormalized
+}
+
+interface NavigationError extends RouterError {
+  type:
+    | ErrorTypes.NAVIGATION_GUARD_REDIRECT
+    | ErrorTypes.NAVIGATION_ABORTED
+    | ErrorTypes.NAVIGATION_CANCELLED
+  from: RouteLocationNormalized
+  to: RouteLocation | RouteLocationNormalized
```
Comment (posva) [inline src/errors-new.ts]: Only the NavigationRedirect has this limitation. We should have an extra type that extends Navigation Error for that one

**Pull request merged**